refactor: [ANDROSDK-1736] reserved values from RxJava to Coroutines

dan-dhis2 opened the pull request:
refactor:
- All RxJava calls to Coroutines for ReservedValues
fix:
- EndpointCall.kt extending

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- ConstantModule related classes to Kotlin

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- ConstantModule related classes to Kotlin

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- ReservedValues network call to coroutines

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- ReservedValues network call to coroutines

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- ConstantModule network call to coroutines

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- ConstantModule network call to coroutines

andresmr committed:
[ANDROSDK-1742] Update paging library to 3.2.0

andresmr committed:
[ANDROSDK-1742] update tests

andresmr committed:
[ANDROSDK-1742] add new method getPagingData into ReadOnlyCollectionRepository

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- ConstantModule calls from RxJava to Coroutines
fix:
- AuthorityEndpointCallFactory to access Kotlin properties
style:
- Added line to Eof

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- ConstantModule calls from RxJava to Coroutines
fix:
- AuthorityEndpointCallFactory to access Kotlin properties
style:
- Added line to Eof

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- Cleaned up some nullables
- Simplify fetcher suspend call

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- Cleaned up some nullables
- Simplify fetcher suspend call

andresmr committed:
[ANDROSDK-1742] Update PagingSourceRepository implementations

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- TypedModuleDownloader.kt to coroutines
- MetadataCallShould.kt to handle new calls
- MetadataCall.kt to handle coroutines for those that depend on TypedModuleDownloader.kt
style:
- Minor styling fixes

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- TypedModuleDownloader.kt to coroutines
- MetadataCallShould.kt to handle new calls
- MetadataCall.kt to handle coroutines for those that depend on TypedModuleDownloader.kt
style:
- Minor styling fixes

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- Make interfaces functional

dan-dhis2 committed:
[ANDROSDK-1732]
refactor:
- Remove unnecessary "runBlocking" calls from MetadataCall.kt

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- All RxJava calls to Coroutines for ReservedValues
fix:
- EndpointCall.kt extending

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- All RxJava calls to Coroutines for ReservedValues
fix:
- EndpointCall.kt extending

dan-dhis2 committed:
Rename .java to .kt

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- EndpointCall to UidsCallFactoryImpl and ListCallFactoryImpl to Kotlin
- Remove deprecated usages of BaseQuery in favor of BaseQueryKt
- Adapt BaseQueryKt to Java call from AutoValue class

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- EndpointCall to UidsCallFactoryImpl and ListCallFactoryImpl to Kotlin
- Remove deprecated usages of BaseQuery in favor of BaseQueryKt
- Adapt BaseQueryKt to Java call from AutoValue class

dan-dhis2 committed:
Rename .java to .kt

vgarciabnz committed:
Merge pull request #2017 from dhis2/ANDROSDK-1732_replace_rxjava_constantmodule

refactor: [ANDROSDK-1732] replace RxJava for ConstantModule

vgarciabnz commented on core/src/main/java/org/hisp/dhis/android/core/arch/call/fetchers/internal/ListNoResourceWithErrorCatcherCallFetcher.kt:
```diff
@@ -25,30 +25,20 @@
-        return apiCallExecutor.executeObjectCallWithErrorCatcher(getCall(), errorCatcher);
+    @Throws(D2Error::class)
+    suspend fun fetch(): List<P> {
+        return coroutineAPICallExecutor.wrap(errorCatcher = errorCatcher) { call }.getOrThrow()
```
This `call` isn't a function, it is a val, I am not sure it would work fine. Anyway, this class is never used, it can be removed.

vgarciabnz commented on core/src/main/java/org/hisp/dhis/android/core/trackedentity/internal/TrackedEntityAttributeReservedValueService.kt:
```diff
@@ -25,28 +25,28 @@
-            @Path(TRACKED_ENTITY_ATTRIBUTE_UID) String trackedEntityAttributeUid,
-            @Query("numberToReserve") Integer numberToReserve);
+    suspend fun generateAndReserve(
+        @Path(TRACKED_ENTITY_ATTRIBUTE_UID) trackedEntityAttributeUid: String?,
```
All these parameters could be NonNull

vgarciabnz commented on core/src/main/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManager.kt:
```diff
@@ -311,7 +308,7 @@ class TrackedEntityAttributeReservedValueManager @Inject internal constructor(
         if (remainingValues < minNumberToTryFill) {
             val numberToReserve = fillUpTo!! - remainingValues
             downloadValues(
-                attribute, organisationUnit, numberToReserve, pattern, storeError
```
The purpose of this storeError was to optionally store the error based on the method that originated the query: if the query is generated as a side effect of `getValue()`, it is fine if the query fails because the device can be offline and the user didn't trigger the query. If the query is generated explicitly by the user, it is important to store the error.

The `storeError` property can be added to the TrackedEntittyAttributeReservedValueQuery and then applied in the coroutineExecutor.

vgarciabnz reviewed (changes requested):
This PR has conflicting files

vgarciabnz committed:
Merge pull request #2016 from dhis2/ANDROSDK-1742

[ANDROSDK-1742] Update pagin library to Paging3

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- Use of ListNoResourceWithErrorCatcherCallFetcher.kt
- Remove unneeded nullables
- (re)added 'storeError'

dan-dhis2 committed:
[ANDROSDK-1736]
refactor:
- Use of ListNoResourceWithErrorCatcherCallFetcher.kt
- Remove unneeded nullables
- (re)added 'storeError'

vgarciabnz commented on core/src/main/java/org/hisp/dhis/android/core/arch/call/factories/internal/QueryCoroutineCallFactoryImpl.kt:
```diff
@@ -34,9 +34,9 @@ import org.hisp.dhis.android.core.arch.call.queries.internal.BaseQueryKt
 ) : QueryCallFactory<P, Q> {
-    override suspend fun create(query: Q): MutableList<P> {
+    override suspend fun create(query: Q): List<P> {
         val objects: List<P> = fetcher(query)
```
Just for convention, it should be `fetcher(query).fetch()` instead of calling the `fetch()` method inside the `fetcher()`.

vgarciabnz commented on core/src/main/java/org/hisp/dhis/android/core/arch/call/factories/internal/QueryCoroutineCallFactoryImpl.kt:
```diff
@@ -34,9 +34,9 @@ import org.hisp.dhis.android.core.arch.call.queries.internal.BaseQueryKt
+    override suspend fun create(query: Q): List<P> {
         val objects: List<P> = fetcher(query)
         processor(query).process(objects)
         return objects.toMutableList()
```
This `toMutableList()` is not needed, isn't it?

dan-dhis2 committed:
[ANDROSDK-1736]
style:
- Post-rebase code styling

dan-dhis2 committed:
Merge remote-tracking branch 'origin/ANDROSDK-1736_refactor_reserved_values_call' into ANDROSDK-1736_refactor_reserved_values_call

# Conflicts:
#	core/src/androidTest/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManagerRealIntegrationShould.kt
#	core/src/main/java/org/hisp/dhis/android/core/arch/call/factories/internal/ListCallFactoryImpl.kt
#	core/src/main/java/org/hisp/dhis/android/core/arch/call/factories/internal/UidsCallFactoryImpl.kt
#	core/src/main/java/org/hisp/dhis/android/core/arch/call/fetchers/internal/ListNoResourceWithErrorCatcherCallFetcher.kt
#	core/src/main/java/org/hisp/dhis/android/core/arch/call/internal/EndpointCall.kt
#	core/src/main/java/org/hisp/dhis/android/core/arch/call/queries/internal/BaseQueryKt.kt
#	core/src/main/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManager.kt
#	core/src/main/java/org/hisp/dhis/android/core/trackedentity/internal/TrackedEntityAttributeReservedValueEndpointCallFactory.kt
#	core/src/main/java/org/hisp/dhis/android/core/trackedentity/internal/TrackedEntityAttributeReservedValueService.kt

dan-dhis2 commented:
Closing this PRs due to messed git history
